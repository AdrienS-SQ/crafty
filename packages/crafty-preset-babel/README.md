<table>
<tr><th>Compatible Runners</th><td>

- [Gulp](05_Packages/02_crafty-runner-gulp.md)
- [rollup.js](05_Packages/02_crafty-runner-rollup.md)
- [Webpack](05_Packages/02_crafty-runner-webpack.md)

</td></tr>
<tr><th>Test Runners</th><td>

- [Jest](05_Packages/05_crafty-preset-jest.md)

</td></tr>
<tr><th>Linters</th><td>

Provides ESLint, configured with [`eslint-plugin-swissquote`](05_Packages/10_eslint-plugin-swissquote.md)

</td></tr>
<tr><th>Commands</th><td>

- `jsLint`: Lint JavaScript files, this is a facade for ESLint, pre-configured with our preset.

</td></tr>
</table>

[TOC]

## Description

**Babel** is the leading tool to compile EcmaScript 2015+ to EcmaScript 5, combined with **ESLint** to lint your code, you get the best preset to get started.

## Features

`@swissquote/crafty-preset-babel` is able to configure **Babel** with **Webpack** and **rollup.js**. This preset also supports **Gulp** but in this case it concatenates and minifies the files, it doesn't resolve imports.

[Our Babel preset](05_Packages/10_babel-preset-swissquote.md)

[Read more](./JavaScript_Features.md)

## Linting

In `@swissquote/crafty-preset-babel` JavaScript is linted with **ESLint**, a powerful linter that supports plugins, our configuration follows the Swissquote JavaScript Guideline.

[Read more](./JavaScript_Linting.md)

## Installation

```bash
npm install @swissquote/crafty-preset-babel --save
```

```javascript
module.exports = {
  presets: [
    "@swissquote/crafty-preset-babel",
    "@swissquote/crafty-runner-webpack", // optional
    "@swissquote/crafty-runner-gulp" // optional
  ],
  js: {
    app: {
      runner: "webpack", // Webpack, Gulp or rollup.js (optional if you have a single runner defined)
      source: "js/app.js"
    }
  }
};
```

## Usage

### With Webpack / rollup.js

Both runners have the same features in regards to Babel support.
They will resolve your modules recursively and bundle them in one file or more if you do some code-splitting.

#### JavaScript External assets

By default, all bundlers include all external dependencies in the final bundle, this works fine for applications, but if you wish to build a multi-tenant application or a library, you don't wish to include all dependencies, because you'll end up with the same dependency duplicated.

The `externals` option allows you to define a list of libraries that are provided and should not be embedded in the build, here is an example :

```javascript
module.exports = {
    ...
    // patterns are strings or globs
    externals: ["react", "react-dom", "squp", "squp/**"],
    ...
    js: {
        app: {
            // You can provide more items here, they will be merged with the main list for this bundle
            externals: ["my-plugin", "my-plugin/**"]
            ...
        }
    }
    ...
}
```

In this example `react`, `react-dom` and all modules starting with `squp/` will be treated as external

> You can see that globs were used here, note that they work for Webpack but rollup.js needs complete strings.

### With Gulp

Gulp will not bundle your files like Webpack and rollup.js do, instead it will generate one output file per input file.
This is useful if you are creating a library as it's the role of the final application to tree-shake what it doesn't need from your library.

Tree-shaking is powerful but is sub-optimal on big files as some code patterns are recognized as side-effects and thus aren't removed from your bundle even if they aren't used.

## Usage with Jest

If you include both `crafty-preset-babel` and `crafty-preset-jest`.
When running your tests with `crafty test` this preset will be use to convert all `.js` and `.jsx` files (source and test files)

## Configuration

### Bundle options

| Option   | Type    | Optional ? | Runner | Description                                                                                                                     |
| -------- | ------- | ---------- | ------ | ------------------------------------------------------------------------------------------------------------------------------- |
| `concat` | Boolean | Yes        | Gulp   | This will merge all files together, outputting a single file. (This doesn't resolve imports, use Webpack or rollup.js for this) |

### Adding Babel plugins and presets

You can add, replace or remove plugins and add options to our default Babel configuration.
To see which plugins are already included, you can go to the [Swissquote Preset for Babel](05_Packages/10_babel-preset-swissquote.md) page.

```javascript
module.exports = {
  /**
   * Represents the extension point for Babel configuration
   * @param {Crafty} crafty - The instance of Crafty.
   * @param {Object} bundle - The bundle that is being prepared for build (name, input, source, destination)
   * @param {Object} babelConfig - The current Babel configuration
   */
  babel(crafty, bundle, babelConfig) {
    babelConfig.plugins.push(require.resolve("@babel/plugin-transform-property-literals"));
  }
};
```

After you did `npm install --save-dev babel-plugin-transform-es5-property-mutators` before, Babel will now use this plugin as well in each run.

This method is called once per bundle, so you can customize each bundle's configuration differently.

### Linting options

You can read about the linting options in the page about [Read more](./JavaScript_Linting.md)

## Commands

### `crafty jsLint`

This linter will leverage ESLint to lint your JavaScript files with the Swissquote presets pre-configured. All [ESLint CLI](https://eslint.org/docs/user-guide/command-line-interface) options are valid here.

The additions made by this command are:

- Pre-configured rules, defined by [`eslint-plugin-swissquote`](05_Packages/10_eslint-plugin-swissquote.md) activated using `--preset`.
- Uses `babel-eslint` as a parser to support new syntax that ESLint doesn't understand yet.

there are 4 presets available for you :

- `format` Base formatting rules, should work on any code (included in `legacy` and `recommended`)
- `node` Adds environment information for Node.js
- `legacy` For all your EcmaScript 5 code
- `recommended` For al your EcmaScript 2015+ code, also contains rules for React

Setting presets is done with the `--preset` option

The order of the presets is important as some rules might override previous ones.

For example:

```bash
crafty jsLint src/** --preset format --preset node --preset recommended
```

If no preset is specified `recommended` is used.

If you pass the `--fix` flag it will fix all the errors it can and write them directly to the file.
