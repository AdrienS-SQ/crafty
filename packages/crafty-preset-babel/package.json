{
  "name": "@swissquote/crafty-preset-babel",
  "version": "1.3.0",
  "license": "Apache-2.0",
  "repository": "https://github.com/swissquote/crafty.git",
  "publishConfig": {
    "access": "public"
  },
  "main": "src/index.js",
  "dependencies": {
    "@babel/core": "7.1.2",
    "@babel/runtime": "7.1.2",
    "@swissquote/babel-preset-swissquote": "1.3.0",
    "@swissquote/eslint-plugin-swissquote": "1.3.0",
    "babel-core": "7.0.0-bridge.0",
    "babel-loader": "8.0.4",
    "babel-plugin-istanbul": "5.0.1",
    "babel-preset-jest": "23.2.0",
    "debug": "^4.0.0",
    "eslint": "5.6.1",
    "eslint-loader": "2.1.1",
    "gulp-babel": "8.0.0",
    "gulp-concat": "2.6.1",
    "gulp-eslint": "5.0.0",
    "gulp-newer": "1.4.0",
    "gulp-sourcemaps": "2.6.4",
    "gulp-uglify": "3.0.1",
    "merge": "1.2.0",
    "resolve-from": "4.0.0",
    "rollup-plugin-babel": "4.0.3",
    "rollup-plugin-eslint": "5.0.0",
    "tmp": "0.0.33",
    "uglify-es": "3.3.9"
  },
  "engines": {
    "node": ">=8.0.0"
  }
}
