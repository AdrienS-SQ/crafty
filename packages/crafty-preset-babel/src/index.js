const merge = require("merge");
const tmp = require("tmp");
const fs = require("fs");
const path = require("path");
const rollupBabel = require("rollup-plugin-babel");
const rollupEslint = require("rollup-plugin-eslint");

const babelConfigurator = require("./babel");
const createTask = require("./gulp");

const MODULES = path.join(__dirname, "..", "node_modules");

module.exports = {
  defaultConfig(config) {
    // For some shitty reason, eslint.CLIEngine doesn't support "extends",
    // it has to be done through a configuration file
    const esConfig = tmp.fileSync({
      prefix: "sq-gulp-config-",
      postfix: ".json"
    }).name;

    fs.writeFileSync(
      esConfig,
      JSON.stringify({ extends: ["plugin:@swissquote/swissquote/recommended"] })
    );

    return {
      bundleTypes: { js: "js" },

      // In case of browser support changes, we prefer having a fixed version
      eslintReactVersion: "15.0",

      // ESLint Override Rules
      eslint: {
        useEslintrc: false,
        plugins: ["@swissquote/swissquote"],
        configFile: esConfig
      },

      // Pass options to uglifyJS, used by both webpack and gulp-uglify
      uglifyJS: { compress: true, sourceMap: true }
    };
  },
  config(config) {
    // Add eslint react version
    config.eslint = merge.recursive(config.eslint, {
      baseConfig: {
        settings: { react: { version: config.eslintReactVersion } }
      }
    });

    return config;
  },
  jest(crafty, options) {
    options.moduleDirectories.push(MODULES);
    options.transform["\\.(js|jsx)$"] = require.resolve("./jest-transformer");
    options.moduleFileExtensions.push("jsx");

    options.globals.BABEL_OPTIONS = babelConfigurator(crafty, "test", {});
  },
  commands() {
    return {
      jsLint: {
        //eslint-disable-next-line no-unused-vars
        command: function(crafty, input, cli) {
          global.craftyConfig = crafty.config;
          require("./commands/jsLint");
        },
        description: "Lint JavaScript for errors"
      }
    };
  },
  bundleCreator(crafty) {
    const configurators = { js: {} };

    if (
      crafty.config.loadedPresets.some(
        preset => preset.presetName === "@swissquote/crafty-runner-gulp"
      )
    ) {
      configurators.js["gulp/babel"] = (
        crafty,
        bundle,
        gulp,
        StreamHandler
      ) => {
        gulp.task(bundle.taskName, createTask(crafty, bundle, StreamHandler));
        crafty.watcher.add(bundle.watch || bundle.source, bundle.taskName);
      };
    }

    return configurators;
  },
  rollup(crafty, bundle, rollupConfig) {
    // Add Eslint configuration
    //TODO :: throw error after all files are linted, not after first error
    rollupConfig.input.plugins.eslint = {
      plugin: rollupEslint,
      weight: -20,
      options: Object.assign({}, crafty.config.eslint, {
        throwOnError: crafty.getEnvironment() === "production",
        exclude: ["node_modules/**"],
        include: ["**/*.js", "**/*.jsx"]
      })
    };

    const babelOptions = babelConfigurator(
      crafty,
      crafty.getEnvironment() === "production" ? "production" : "development",
      bundle
    );

    // Make it that babel references helpers instead of embedding them
    // this allows rollup to de-duplicate them.
    babelOptions.plugins.push(require.resolve("babel-plugin-external-helpers"));

    // Webpack handles this at the loader level, but Rollup needs it this way.
    babelOptions.exclude = ["node_modules/**"];
    babelOptions.include = ["**/*.js", "**/*.jsx"];

    rollupConfig.input.plugins.babel = {
      plugin: rollupBabel,
      weight: 20,
      options: babelOptions
    };
  },
  webpack(crafty, bundle, chain) {
    chain.resolve.extensions.add(".jsx");
    chain.resolve.modules.add(MODULES);
    chain.resolveLoader.modules.add(MODULES);

    const options = babelConfigurator(
      crafty,
      crafty.getEnvironment() === "production" ? "production" : "development",
      bundle
    );

    if (crafty.getEnvironment() === "production") {
      options.cacheDirectory = true;
    }

    // EcmaScript 2015+
    chain.module
      .rule("babel")
      .test(/\.jsx?$/)
      .exclude.add(/(node_modules|bower_components)/)
      .end()
      .use("babel")
      .loader("babel-loader")
      .options(options);

    // JavaScript linting
    chain.module
      .rule("lint-js")
      .pre()
      .test(/\.jsx?$/)
      .exclude.add(/(node_modules|bower_components|sq-gulp)/)
      .end()
      .use("eslint")
      .loader("eslint-loader")
      .options(crafty.config.eslint);
  }
};
