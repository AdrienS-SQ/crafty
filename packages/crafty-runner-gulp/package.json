{
  "name": "@swissquote/crafty-runner-gulp",
  "version": "1.0.3",
  "license": "UNLICENSED",
  "repository": "https://github.com/swissquote/crafty.git",
  "main": "./src/index.js",
  "scripts": {
    "build": "tsc",
    "build:watch": "tsc --watch"
  },
  "dependencies": {
    "@swissquote/crafty": "1.0.3",
    "debug": "^3.0.1",
    "glob-watcher": "4.0.0",
    "gulp-plumber": "1.1.0",
    "vinyl-fs": "^2.0.0"
  },
  "engines": {
    "node": ">=4.3.0 <5.0.0 || >=5.10"
  }
}
