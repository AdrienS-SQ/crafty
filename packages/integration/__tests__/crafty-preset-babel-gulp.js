/* global describe, it, expect */

const fs = require("fs");
const path = require("path");

const rimraf = require("rimraf");
const configuration = require("@swissquote/crafty/src/configuration");
const getCommands = require("@swissquote/crafty/src/commands/index");

const testUtils = require("../utils");

const getCrafty = configuration.getCrafty;

it("Loads crafty-preset-babel and does not register gulp tasks", () => {
  const crafty = getCrafty(["@swissquote/crafty-preset-babel"], {});

  const loadedPresets = [
    require("@swissquote/crafty-preset-babel"),
    { presetName: "crafty.config.js" }
  ];

  expect(crafty.config.loadedPresets).toEqual(loadedPresets);

  const commands = getCommands(crafty);
  expect(Object.keys(commands)).toContain("jsLint");

  crafty.createTasks();
  expect(Object.keys(crafty.undertaker._registry.tasks())).toEqual([]);
});

it("Loads crafty-preset-babel, crafty-runner-gulp and registers gulp task", () => {
  const config = { js: { myBundle: { source: "css/style.scss" } } };
  const crafty = getCrafty(
    ["@swissquote/crafty-preset-babel", "@swissquote/crafty-runner-gulp"],
    config
  );

  const loadedPresets = [
    require("@swissquote/crafty-preset-babel"),
    require("@swissquote/crafty-runner-gulp"),
    Object.assign({ presetName: "crafty.config.js" }, config)
  ];

  expect(crafty.config.loadedPresets).toEqual(loadedPresets);

  const commands = getCommands(crafty);
  expect(Object.keys(commands)).toContain("jsLint");

  crafty.createTasks();
  expect(Object.keys(crafty.undertaker._registry.tasks())).toEqual([
    "js_myBundle",
    "js",
    "default"
  ]);
});

it("Compiles JavaScript", () => {
  process.chdir(
    path.join(__dirname, "../fixtures/crafty-preset-babel-gulp/compiles")
  );
  rimraf.sync("dist");

  const result = testUtils.run(["run", "default"]);

  expect(result).toMatchSnapshot();

  expect(fs.existsSync("dist/js/myBundle.min.js")).toBeFalsy();
  expect(fs.existsSync("dist/js/myBundle.min.js.map")).toBeFalsy();

  expect(fs.existsSync("dist/js/script.js")).toBeTruthy();
  expect(fs.existsSync("dist/js/script.js.map")).toBeTruthy();

  expect(fs.existsSync("dist/js/otherfile.js")).toBeTruthy();
  expect(fs.existsSync("dist/js/otherfile.js.map")).toBeTruthy();

  expect(
    fs.readFileSync("dist/js/script.js").toString("utf8")
  ).toMatchSnapshot();
  expect(
    fs.readFileSync("dist/js/otherfile.js").toString("utf8")
  ).toMatchSnapshot();
});

it("Fails gracefully on broken markup", () => {
  process.chdir(
    path.join(__dirname, "../fixtures/crafty-preset-babel-gulp/fails")
  );
  rimraf.sync("dist");

  const result = testUtils.run(["run", "default"]);

  expect(result).toMatchSnapshot();

  expect(fs.existsSync("dist/js/myBundle.min.js")).toBeFalsy();
  expect(fs.existsSync("dist/js/myBundle.min.js.map")).toBeFalsy();
});

it("Compiles JavaScript with custom babel plugin", () => {
  process.chdir(
    path.join(
      __dirname,
      "../fixtures/crafty-preset-babel-gulp/compiles-babel-plugin"
    )
  );
  rimraf.sync("dist");

  const result = testUtils.run(["run", "default"]);

  expect(result).toMatchSnapshot();

  expect(fs.existsSync("dist/js/myBundle.min.js")).toBeFalsy();
  expect(fs.existsSync("dist/js/myBundle.min.js.map")).toBeFalsy();

  expect(fs.existsSync("dist/js/script.js")).toBeTruthy();
  expect(fs.existsSync("dist/js/script.js.map")).toBeTruthy();

  expect(
    fs.readFileSync("dist/js/script.js").toString("utf8")
  ).toMatchSnapshot();
});

it("Compiles JavaScript and concatenates", () => {
  process.chdir(
    path.join(__dirname, "../fixtures/crafty-preset-babel-gulp/concatenates")
  );
  rimraf.sync("dist");

  const result = testUtils.run(["run", "default"]);

  expect(result).toMatchSnapshot();

  expect(fs.existsSync("dist/js/myBundle.min.js")).toBeTruthy();
  expect(fs.existsSync("dist/js/myBundle.min.js.map")).toBeTruthy();

  expect(fs.existsSync("dist/js/script.js")).toBeFalsy();
  expect(fs.existsSync("dist/js/script.js.map")).toBeFalsy();

  expect(fs.existsSync("dist/js/otherfile.js")).toBeFalsy();
  expect(fs.existsSync("dist/js/otherfile.js.map")).toBeFalsy();

  expect(
    fs.readFileSync("dist/js/myBundle.min.js").toString("utf8")
  ).toMatchSnapshot();
});

it("Lints JavaScript", () => {
  process.chdir(
    path.join(__dirname, "../fixtures/crafty-preset-babel-gulp/lints-es5")
  );
  rimraf.sync("dist");

  const result = testUtils.run(["run", "default"]);

  expect(result).toMatchSnapshot();

  // Files aren't generated on failed lint
  expect(fs.existsSync("dist/js/myBundle.min.js")).toBeFalsy();
  expect(fs.existsSync("dist/js/myBundle.min.js.map")).toBeFalsy();
});

it("Lints JavaScript, doesn't fail in development", () => {
  process.chdir(
    path.join(__dirname, "../fixtures/crafty-preset-babel-gulp/lints-es5-dev")
  );
  rimraf.sync("dist");

  const result = testUtils.run(["run", "default"]);

  expect(result).toMatchSnapshot();

  // Files aren't generated on failed lint
  expect(fs.existsSync("dist/js/myBundle.min.js")).toBeFalsy();
  expect(fs.existsSync("dist/js/myBundle.min.js.map")).toBeFalsy();
});
