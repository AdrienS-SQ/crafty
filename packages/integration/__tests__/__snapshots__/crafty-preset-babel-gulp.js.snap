// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compiles JavaScript 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "
[__:__:__] Starting Crafty __version__...
[__:__:__] Files will be stored in:
            js: __PATH__/packages/integration/fixtures/crafty-preset-babel-gulp/compiles/dist/js
[__:__:__] Starting 'default' ...
[__:__:__] Starting 'js' ...
[__:__:__] Starting 'js_myBundle' ...
[__:__:__] Finished 'js_myBundle' after ____ ms
[__:__:__] Finished 'js' after ____ ms
[__:__:__] Finished 'default' after ____ ms
",
}
`;

exports[`Compiles JavaScript 2`] = `
"function _classCallCheck(e,n){if(!(e instanceof n))throw new TypeError(\\"Cannot call a class as a function\\")}var _createClass=function(){function e(e,n){for(var t=0;t<n.length;t++){var a=n[t];a.enumerable=a.enumerable||!1,a.configurable=!0,\\"value\\"in a&&(a.writable=!0),Object.defineProperty(e,a.key,a)}}return function(n,t,a){return t&&e(n.prototype,t),a&&e(n,a),n}}(),Test=function(){function e(n){_classCallCheck(this,e),this.name=n}return _createClass(e,[{key:\\"sayHi\\",value:function(){console.log(\\"Hello \\"+this.name+\\" !\\")}}]),e}(),test=new Test(\\"Stéphane\\");test.sayHi();
//# sourceMappingURL=script.js.map
"
`;

exports[`Compiles JavaScript 3`] = `
"setTimeout(function(){console.log(\\"Welcome to this website\\")},1e3);
//# sourceMappingURL=otherfile.js.map
"
`;

exports[`Compiles JavaScript and concatenates 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "
[__:__:__] Starting Crafty __version__...
[__:__:__] Files will be stored in:
            js: __PATH__/packages/integration/fixtures/crafty-preset-babel-gulp/concatenates/dist/js
[__:__:__] Starting 'default' ...
[__:__:__] Starting 'js' ...
[__:__:__] Starting 'js_myBundle' ...
[__:__:__] Finished 'js_myBundle' after ____ ms
[__:__:__] Finished 'js' after ____ ms
[__:__:__] Finished 'default' after ____ ms
",
}
`;

exports[`Compiles JavaScript and concatenates 2`] = `
"function _classCallCheck(e,n){if(!(e instanceof n))throw new TypeError(\\"Cannot call a class as a function\\")}setTimeout(function(){console.log(\\"Welcome to this website\\")},1e3);var _createClass=function(){function e(e,n){for(var t=0;t<n.length;t++){var a=n[t];a.enumerable=a.enumerable||!1,a.configurable=!0,\\"value\\"in a&&(a.writable=!0),Object.defineProperty(e,a.key,a)}}return function(n,t,a){return t&&e(n.prototype,t),a&&e(n,a),n}}(),Test=function(){function e(n){_classCallCheck(this,e),this.name=n}return _createClass(e,[{key:\\"sayHi\\",value:function(){console.log(\\"Hello \\"+this.name+\\" !\\")}}]),e}(),test=new Test(\\"Stéphane\\");test.sayHi();
//# sourceMappingURL=myBundle.min.js.map
"
`;

exports[`Compiles JavaScript with custom babel plugin 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "
[__:__:__] Starting Crafty __version__...
[__:__:__] Files will be stored in:
            js: __PATH__/packages/integration/fixtures/crafty-preset-babel-gulp/compiles-babel-plugin/dist/js
[__:__:__] Starting 'default' ...
[__:__:__] Starting 'js' ...
[__:__:__] Starting 'js_myBundle' ...
[__:__:__] Finished 'js_myBundle' after ____ ms
[__:__:__] Finished 'js' after ____ ms
[__:__:__] Finished 'default' after ____ ms
",
}
`;

exports[`Compiles JavaScript with custom babel plugin 2`] = `
"var foo = {
  \\"catch\\": function _catch() {}
};

//eslint-disable-next-line no-console
console.log(foo.catch);
//# sourceMappingURL=script.js.map
"
`;

exports[`Fails gracefully on broken markup 1`] = `
Object {
  "status": 1,
  "stderr": "[__:__:__] 'js_myBundle' errored after ____ ms
[__:__:__] SyntaxError in plugin 'gulp-babel'
Message:
    __PATH__/packages/integration/fixtures/crafty-preset-babel-gulp/fails/js/script.js: Unexpected token (14:6)
[__:__:__] 'js' errored after ____ ms
[__:__:__] 'default' errored after ____ ms

events.js:182
      throw er; // Unhandled 'error' event
      ^
ESLintError: Failed with 1 error
",
  "stdout": "
[__:__:__] Starting Crafty __version__...
[__:__:__] Files will be stored in:
            js: __PATH__/packages/integration/fixtures/crafty-preset-babel-gulp/fails/dist/js
[__:__:__] Starting 'default' ...
[__:__:__] Starting 'js' ...
[__:__:__] Starting 'js_myBundle' ...
[__:__:__]
__PATH__/packages/integration/fixtures/crafty-preset-babel-gulp/fails/js/script.js
  14:7  error  Parsing error: Unexpected token

  12 | }
  13 |
> 14 | const test = new Test(\\"Stéphane\\");
     |       ^
  15 | test.sayHi();
  16 |

✖ 1 problem (1 error, 0 warnings)

",
}
`;

exports[`Lints JavaScript 1`] = `
Object {
  "status": 1,
  "stderr": "[__:__:__] 'js_myBundle' errored after ____ ms
[__:__:__] ESLintError in plugin 'gulp-eslint'
Message:
    Failed with 4 errors
Details:
    domainEmitter: [object Object]
    domain: [object Object]
    domainThrown: false
[__:__:__] 'js' errored after ____ ms
[__:__:__] 'default' errored after ____ ms
",
  "stdout": "
[__:__:__] Starting Crafty __version__...
[__:__:__] Files will be stored in:
            js: __PATH__/packages/integration/fixtures/crafty-preset-babel-gulp/lints-es5/dist/js
[__:__:__] Starting 'default' ...
[__:__:__] Starting 'js' ...
[__:__:__] Starting 'js_myBundle' ...
[__:__:__]
__PATH__/packages/integration/fixtures/crafty-preset-babel-gulp/lints-es5/js/script.js
  3:19  error  A space is required after ','                                                  comma-spacing
  4:1   error  Expected indentation of 2 spaces but found 3                                   indent
  8:1   error  Opening curly brace does not appear on the same line as controlling statement  brace-style
  9:12  error  Expected '===' and instead saw '=='                                            eqeqeq

✖ 4 problems (4 errors, 0 warnings)
  3 errors, 0 warnings potentially fixable with the \`--fix\` option.

",
}
`;

exports[`Lints JavaScript, doesn't fail in development 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "
[__:__:__] Starting Crafty __version__...
[__:__:__] Files will be stored in:
            js: __PATH__/packages/integration/fixtures/crafty-preset-babel-gulp/lints-es5-dev/dist/js
[__:__:__] Starting 'default' ...
[__:__:__] Starting 'js' ...
[__:__:__] Starting 'js_myBundle' ...
[__:__:__]
__PATH__/packages/integration/fixtures/crafty-preset-babel-gulp/lints-es5-dev/js/script.js
  3:19  warning  A space is required after ','                                                  comma-spacing
  4:1   warning  Expected indentation of 2 spaces but found 3                                   indent
  8:1   warning  Opening curly brace does not appear on the same line as controlling statement  brace-style

✖ 3 problems (0 errors, 3 warnings)
  0 errors, 3 warnings potentially fixable with the \`--fix\` option.

[__:__:__] Finished 'js_myBundle' after ____ ms
[__:__:__] Finished 'js' after ____ ms
[__:__:__] Finished 'default' after ____ ms
",
}
`;
